manpages/patch2imod.man
.na
.nh
.TH patch2imod 1 2.42 BL3DEMC
.SH NAME
patch2imod \- makes model from patch displacement, warping, or residual file
.SH SYNOPSIS
patch2imod  [options]  patch_file  IMOD_model_file
.SH DESCRIPTION
patch2imod creates an IMOD model from a file containing 3-D displacements at
a series of positions, such as the patch file generated by corrsearch3d(1).
Each displacement will be represented by a vector starting at the position
of the patch, and whose magnitude and direction are determined by the 3-D
displacement multiplied by a scale factor.  These vectors are simply open
contours consisting
of two points.  In addition, if the patch file contains correlation
coefficients or residual values, these values will be stored in the model.
.P
The resulting model can be viewed to determine the location of patches with 
aberrant displacements and even edited to remove such patches.  The
easiest way to do this is to start 3dmod with just the model file and no
image file, open the model view window, and use the right mouse button in
the model view window to select vectors that differ greatly from 
their neighbors.  The currently selected vector(s) will be highlighted by
being thicker.
They may then be deleted with the D hot key.  The edited
model can be converted back to a patch file with imod2patch(1).
.P
The correlation or residual values can be visualized by opening the "Values"
panel in the model view Object Edit window and selecting "Show stored values".
The color display can be manipulated and subsets of vectors can be selected.
.P
If the patch file contains a column of numbers after the correlation
coefficient or residual values, these will
also be stored in the model but cannot be visualized.  They will be
output to a new patch file by imod2patch(1). 
.P
One can also
read the model in on top of a tomogram; by default it is marked as flipped so
that it will display correctly on tomograms in their original orientations.
.P
The scaling is stored as the pixel size in the model header.  Do not change
this value, or imod2patch will not be able to convert the displacements back
correctly.
.P
The program will also convert a list of 2-D residuals produced by Tiltalign(1).
In this case, the model is not marked as flipped, so that it can be displayed
on top of the original image file.  A file of warping transformations can also
be converted to vectors, and again the model is not marked as flipped.
.SH OPTIONS
.TP
.B -s \fIvalue\fR
Change the vector length scaling to the given value; the default is 10.
.TP
.B -f
Do not flip Y and Z coordinates in the model.  Use this option to display the
model on a tomogram that has been flipped or rotated from its original
orientation.  The option is not needed with a Tiltalign(1) residual file or
with a warping transformation file.
.TP
.B -c \fIsize\fR
Include four clipping planes in the model to enclose a square area of the
given size in the center of the model.  These planes will not be on by default,
but can be turned on and manipulated together in the model view Object Edit
window.  These planes could be helpful for seeing a subset of vectors when the
model is rotated.
.TP
.B -n \fIname>
Give this name to the model object.  The name is useful for indicating what
kind of values are stored with the vectors.
.TP
.B -d
Set the flag to display stored values in false color in 3dmod.  This option has
the same effect as turning on the "Show stored values" button in the Values
panel of the Edit-Objects dialog of the Model View window.
.TP
.B -z
Ignore zero values when computing the mean and SD of values.  The program
computes the actual maximum value and the mean and SD, and then stores a
maximum value that is limited to 10 SDs above the mean.  This stored value
determines the maximum that is shown in the model view object edit window of
3dmod.  This option is needed when there are a very large numbers of zero values.
.TP
.B -l
Read all lines in the file as lines with vectors.  By default, the program
reads the first line to determine the number of lines, then reads that
number of lines as vectors. 
.TP
.B -w
Read the input file as a warping transformation file like those produced by Midas(1)
and Tiltxcorr(1).  Both control point and warping grid files can be
converted.  The negative of the warping vector values will be displayed
because those values are inverse transform values.
.TP
.B -t
Assign a time to each contour, equal to its Z value plus 1.  This option may
be useful when examining vector models from a warping transformation file or
from the edge function output of Blendmont(1), because it would allow one to
examine vectors for one section or edge function at a time in the Model View
window instead of in the Zap window.  One can step through times with the 1
and 2 hot keys, but there may be no good way to select a specific time out of
many.
.SH FILES
Unless the 
.B -l
option or the 
.B -w
option is given, the first line of the file must have the number of vectors in
the file.  If this line contains "residuals", then each succeeding line
should have five values: X, Y, Z, delta X and delta Y.  X and Y may be
floating point values in this case, and Z must be an integer.  Other than the
residual case, each line may have 6 to 8 values, where the first 6 are X, Y,
Z, delta X, delta Y, and delta Z.  X, Y, and Z must all be integers in this
case.  A seventh value is stored at a general value displayable in 3dmod,
and an eighth value is stored as the second type of general value.
.P
If the model file already exists, it becomes a backup file with ~ added to its
name.
.SH AUTHOR
David Mastronarde
.SH SEE ALSO
imod2patch(1), patchcrawl3d(1), patchcorr3d(1), 3dmod(1), 3dmodv(1),
findwarp(1), refinematch(1), tiltalign(1), tiltxcorr(1), midas(1)
.SH BUGS
Email bug reports to mast@colorado.edu.


History
IMOD History
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
12/6/13: Added option to Imodinfo to get contour lengths broken out by
fine-grained color, and made -o option provide object list for all outputs.

12/5/13: Updated the eTomo manpage.  Improved the eTomo help statement.
(Bug# 1764)

12/3/13: In eTomo - PEET, excluding the quoted values functionality.
(Bug# 1631)

11/30/13: Changes to Tiltalign to deal with patch tracking data: it can select
a subset of patches for the global alignment to avoid solving for too many
variables; it can do robust fitting with weights assigned either to individual
points or to whole contours.

11/28/13: Made changes in the search routine used by Tiltalign/Beadtrack to
reduce the high risk of "minimization errors" when starting close to the final
solution; also ignore particular errors after first round of robust fitting.

11/28/13: In eTomo allowing quoted values in autodocs.(Bug# 1631)

11/21/13: IMOD_4-6-33
11/21/13: Changed Corrsearch3d to use dynamic allocation, load from file more 
efficiently, and use multiple threads for some steps.

11/20/13: In eTomo - Final Aligned Stack, prevented the creation of a bad
rotation.xf file when the final aligned image stack is not present.(Bug# 1762)

11/20/13: In eTomo - Serial Sections, fixed the Newstack progress bar.
(Bug# 1749)

11/20/13: Fixed 3dmod crash from loading many files through startup window.

11/20/13: Made multi-line INFO and WARNING messages more parsable in
etomo_err.log.(Bug# 1694)

11/19/13: In eTomo - Combine, prevented the dataset from being damaged when
the Combine dialog is opened too early.(Bug# 1652)

11/19/13: In uitest, fixed a problem with finding the main panel.(Bug# 1761)

11/18/13: In eTomo made multi-line error messages parsable by adding an empty
line after each one.(Bug# 1694)

11/18/13: In eTomo - Coarse Alignment, added support for tiltxcorr parameters
SearchMagChanges and ViewsWithMagChanges.(Bug# 1760)

11/15/13: In eTomo added support for parallel processing using multiple GPUs
per computer.(Bug# 1436)

11/15/13: In eTomo reduced the excessive debug output.(Bug# 1763)

11/13/13: In eTomo - Serial Sections, added a Sobel filter checkbox and
improved the behavior of the alignment panel.(Bug# 1755)

11/11/13: IMOD_4-6-32
11/5/13: Fixed bug that crashed Imodfindbeads when adjusting size downward.

11/4/13: In eTomo - Fiduacial Model dialog, fixed a problem with deleting
autofidseed temporary files in Windows.(Bug# 1759)

11/4/13: Make xfalign apply shift limits in initial cross-correlation.

11/4/13: Fixed scale bar drawing from montage snapshot of model view window.

11/4/13: Made mrc2tif use 4 or 5 digits for all output file numbers if needed.

11/2/13: Switched display of residual model to show arrows.

11/1/13: In eTomo - Serial Sections, improved the context sensitive pop-up
menus.(Bug# 1748)

11/1/13: In eTomo - Fiducial Model, added fields for AdjustSizes and
ClusteredPointsAllowed to the Autofidseed panel.(Bug# 1743)

11/1/13: Made Tiltxcorr able to find mag changes at selected views; for a 
bidirectional series, Copytomocoms puts the option with the view number in
xcorr.com.

11/1/13: Fixed bug in imodtrans -3 option introduced 12/8/12.

10/30/13: IMOD_4-6-31
10/29/13: Fixed a problem in eTomo where the parallel panel did not appear
when eTomo reconnected to a parallel process.  This problem appeared when
Pause and Resume had been previously used.(Bug# 1757)

10/29/13: Fixed the eTomo version string.(Bug# 1744)

10/28/13: In eTomo fixed a bug where stack header data was not getting into
the dataset file when eTomo was run from batchruntomo.(Bug# 1758)

10/27/13: Added option to Tilt to get output in parallel (X/Y) slices without
inverting handedness. 

10/25/13: In eTomo removed dependencies on Cygwin.(Bug# 1713)

10/25/13: Changes in Tiltxcorr: shift limits are now applied in an elliptical
rather than rectangular region; implemented a new, more selective method for
excluding the central peak due to fixed pattern noise; added an option to
compute normalized correlation coefficients at multiple peak positions and
pick the highest (not useful so far).

10/25/13: Changed Beadtrack's elongation file output so that automatic seed
picking would not fail with only 3 beads.

10/24/13: In eTomo - Combine, removed the message complaining about tomograms
A and B not being the same size in X and Y, as this is not an error.
(Bug# 1651)

10/24/13: In eTomo, preventing embedded spaces in dataset names.(Bug# 1667)

10/24/13: Added a antialiasFilter field for the newstack panels in eTomo -
Coarse Alignment and Final Aligned Stack.(Bug# 1737)

10/23/13: In eTomo resizing when the Cleanup dialog is finished.(Bug# 1636)

10/22/13: Fixed another warping-related crash in Midas.

10/22/13: Made it possible to start Zap rubber band from ANY corner.

10/21/13: Fixed a problem that crashed 3dmod when a slicer angle is very
small, such as when doing mouse-rotating.

10/18/13: Added toolbar buttons to Zap and Slicer for drawing a single thick
arrow for snapshots; also added an object property to draw arrowheads at the
ends of contours.

10/17/13: In eTomo - Clean up, added unused, optional image files (such as the
output from mtffilter) to the file chooser.(Bug# 1509)

10/17/13: In eTomo - Clean up, added SIRT .vsr## output files.(Bug# 1533)

10/15/13: Fixed numerical problems in drawing code that crashed Midas on big
images in 64-bit Windows.

10/14/13: Added GiantCriterion, BigDiffCriterion, and ExtraLargeRadius to
eTomo - Preprocessing.(Bug# 1736)

10/9/13: Fixed a problem with running tomodataplots from eTomo under Java 1.7
that prevented the graphs from coming up.(Bug# 1754)

10/9/13: In eTomo fixed a parallel processing problem which prevented the
Resume button from working when the process was pausing.  This problem
appeared in only in reconnected processes.(Bug# 1729)

10/9/13: In eTomo - Nonlinear Anisotropic Diffusion, some test.K... file names

imodutil/patch2imod.c
/*
 *  $Id$
 *
 *  Author: David Mastronarde  email: mast@colorado.edu
 *
 *  Copyright (C) 1995-2004 by Boulder Laboratory for 3-Dimensional Electron
 *  Microscopy of Cells ("BL3DEMC") and the Regents of the University of 
 *  Colorado.  See dist/COPYRIGHT for full copyright notice.
 *
 */

#include <stdio.h>
#include <string.h>
#include "parse_params.h"
#include "imodel.h"
#include "warpfiles.h"

#define P2I_NO_FLIP          1
#define P2I_IGNORE_ZERO      2
#define P2I_COUNT_LINES      4
#define P2I_DISPLAY_VALUES   8
#define P2I_READ_WARP       16
#define P2I_TIMES_FOR_Z     32


#define DEFAULT_SCALE 10.0

static Imod *imod_from_patches(FILE *fin, float scale, int clipSize, char *name, 
                               int flags);

static void usage(char *prog)
{
  imodVersion(prog);
  imodCopyright();
  printf("Usage: %s [options] patch_file output_model\n", prog);
  printf("Options:\n");
  printf("\t-s #\tScale vectors by given value (default %.1f)\n", DEFAULT_SCALE);
  printf("\t-f\tDo NOT flip the Y and Z coordinates\n");
  printf("\t-n name\tAdd given name to model object\n");
  printf("\t-c #\tSet up clipping planes enclosing area of given size\n");
  printf("\t-d #\tSet flag to display values in false color in 3dmod\n");
  printf("\t-z\tIgnore zero values when using SD to limit stored maximum value\n");
  printf("\t-l\tUse all lines in file rather than getting line count from first line\n");
  printf("\t-w\tRead input file as a warping transformation file\n");
  printf("\t-t\tGive each contour a time equal to its Z value plus 1\n");

  exit(1);
}

int main( int argc, char *argv[])
{
  int i;
  FILE *fin = NULL, *fout;
  Imod *Model;
  float scale = 10.0;
  int clipSize = 0;
  int flags = 0;
  int nxWarp, nyWarp, nzWarp, warpFlags, version, indWarp, ibin;
  float warpPixel;
  char *name = NULL;

  /* This name is hard-coded because of the script wrapper needed in Vista */
  char progname[] = "patch2imod";

  if (argc < 2)
    usage(progname);

  setExitPrefix("ERROR: patch2imod - ");
  
  for (i = 1; i < argc ; i++){
    if (argv[i][0] == '-'){
      switch (argv[i][1]){
        
      case 's':
        sscanf(argv[++i], "%f", &scale);
        break;

      case 'n':
        name = strdup(argv[++i]);
        break;

      case 'c':
        clipSize = atoi(argv[++i]);
        break;

      case 'f':
        flags |= P2I_NO_FLIP;
        break;

      case 'z':
        flags |= P2I_IGNORE_ZERO;
        break;

      case 'l':
        flags |= P2I_COUNT_LINES;
        break;

      case 'd':
        flags |= P2I_DISPLAY_VALUES;
        break;

      case 'w':
        flags |= P2I_READ_WARP;
        break;

      case 't':
        flags |= P2I_TIMES_FOR_Z;
        break;

      default:
        exitError("Illegal argument %s", argv[i]);
        break;
      }
    } else
      break;
  }
  if (i > (argc - 2)) {
    printf("ERROR: patch2imod - Wrong # of arguments");
    usage(progname);
  }

  if (flags & P2I_READ_WARP) {
    indWarp = readWarpFile(argv[i++], &nxWarp, &nyWarp, &nzWarp, &ibin, &warpPixel,
                           &version, &warpFlags);
    if (indWarp < 0)
      exitError("Reading %s as a warping file", argv[--i]);
                           
  } else {

    fin = fopen(argv[i++], "r");
    if (!fin)
      exitError("Couldn't open %s", argv[--i]);
  }

  if (imodBackupFile(argv[i])) {
    exitError("Renaming existing output file to %s~", argv[i]);
    exit(1);
  }
  fout = fopen(argv[i], "wb");
  if (!fout)
    exitError("Could not open %s", argv[i]);
  Model = (Imod *)imod_from_patches(fin, scale, clipSize, name, flags);
     
  imodWrite(Model, fout);

  imodDelete(Model);
  exit(0);
}


#define MAXLINE 128

static Imod *imod_from_patches(FILE *fin, float scale, int clipSize, char *name,
                               int flags)
{
  int noflip = flags & P2I_NO_FLIP;
  int ignoreZero = flags & P2I_IGNORE_ZERO;
  int countLines = flags & P2I_COUNT_LINES;
  int displayValues = flags & P2I_DISPLAY_VALUES;
  int readWarp = flags & P2I_READ_WARP;
  int timesForZ = flags & P2I_TIMES_FOR_Z;
  int len, i;
     
  char line[MAXLINE];
  Imod *mod;
  Iobj *obj;
  Istore store, store2;
  Ipoint *pts;
  int ix, iy, iz, itmp;
  int nxWarp, nyWarp, ifControl, nxGrid, nyGrid, numControl, maxProd, izWarp;
  float xStart, yStart, xInterval, yInterval;
  int xmin, ymin, zmin, xmax, ymax, zmax;
  float dx, dy, dz, xx, yy, value, value2, valmin, valmax, tmp;
  double valsum, sumsq, valsd, sdmax;
  float *xVector, *yVector, *xControl, *yControl;
  float *dxGrid = NULL, *dyGrid = NULL;
  int residuals = 0;
  int nvals = 0;
  int dzvary = 0;
  int npatch = 0;
  int nzWarp = 1;
  int nread = 0;
  int contBase = 0;

  if (readWarp) {
    getWarpFileSize(&nxWarp, &nyWarp, &nzWarp, &ifControl);
    maxProd = 0;
    for (iz = 0; iz < nzWarp; iz++) {
      if (ifControl)
        ix = getNumWarpPoints(iz, &numControl);
      else
        ix = getWarpGridSize(iz, &nxGrid, &nyGrid, &numControl);
      if (ix)
        exitError("Getting number of points in warping file at section %d", iz);
      npatch += numControl;
      maxProd = B3DMAX(maxProd, numControl);
    }
    if (!ifControl) {
      dxGrid = B3DMALLOC(float, maxProd);
      dyGrid = B3DMALLOC(float, maxProd);
      if (!dxGrid || !dyGrid)
        exitError("getting memory for warping grids");
    }
  } else if (countLines) {
    while (1) {
      ix = fgetline(fin,line,MAXLINE);
      if (ix > 2)
        npatch++;
      else
        break;
    }
    if (npatch < 1)
      exitError("No usable lines in the file");
    rewind(fin);
  } else {
    fgetline(fin,line,MAXLINE);
    sscanf(line, "%d", &npatch);
    if (npatch < 1)
      exitError("Implausible number of patches = %d.", npatch);

    if (strstr(line, "residuals") != NULL)
      residuals = 1;
  }

  mod = imodNew();     
  if (!mod)
    exitError("Could not get new model");

  if (imodNewObject(mod))
    exitError("Could not get new object");
  obj = mod->obj;
  obj->contsize = npatch;
  obj->cont = imodContoursNew(npatch);
  if (!obj->cont)
    exitError("Could not get contour array");
  obj->flags |= IMOD_OBJFLAG_OPEN;
  if (!residuals && !noflip && !readWarp)
    mod->flags |= IMODF_FLIPYZ;
  if (timesForZ)
    obj->flags |= IMOD_OBJFLAG_TIME;
  mod->pixsize = scale;
  xmin = ymin= zmin = 1000000;
  xmax = ymax = zmax = -1000000;
  valmin = 1.e30;
  valmax = -valmin;
  dz = 0.;
  store.type = GEN_STORE_VALUE1;
  store.flags = GEN_STORE_FLOAT << 2;
  store2.type = GEN_STORE_VALUE2;
  store2.flags = GEN_STORE_FLOAT << 2;

  for (izWarp = 0; izWarp < nzWarp; izWarp++) {
    if (readWarp) {
      iz = izWarp;
      if (ifControl) {
        if (getNumWarpPoints(iz, &npatch) || 
            getWarpPointArrays(iz, &xControl, &yControl, &xVector, &yVector))
          exitError("Getting number of control points or arrays for section %d", iz);
      } else {
        if (getWarpGrid(iz, &nxGrid, &nyGrid, &xStart, &yStart, &xInterval, &yInterval,
                        dxGrid, dyGrid, 0))
          exitError("Getting warp grid for section %d", iz);
        npatch = nxGrid * nyGrid;
      }
    }

    for (i = 0; i < npatch; i++) {
      pts = (Ipoint *)malloc(2 * sizeof(Ipoint));
      if (!pts)
        exitError("Could not get new point array");

      obj->cont[i + contBase].pts = pts;
      obj->cont[i + contBase].psize = 2;
      if (timesForZ)
        obj->cont[i + contBase].time = B3DMAX(1, iz + 1);
      if (readWarp) {
        if (ifControl) {
          xx = xControl[i];
          yy = yControl[i];
          dx = -xVector[i];
          dy = -yVector[i];
        } else {
          ix = i % nxGrid;
          iy = i / nxGrid;
          xx = xStart + ix * xInterval;
          yy = yStart + iy * yInterval;
          dx = -dxGrid[i];
          dy = -dyGrid[i];
        }
      } else {

        len = fgetline(fin,line, MAXLINE);
        if (len < 3)
          exitError("Error reading file at line %d.", i + 1);
        
        /* DNM 7/26/02: read in residuals as real coordinates, without a 
           flip */
        if (residuals) {
          nread = sscanf(line, "%f %f %d %f %f", &xx, &yy, &iz, &dx, &dy);
        } else {
          /* DNM 11/15/01: have to handle either with commas or without,
             depending on whether it was produced by patchcorr3d or 
             patchcrawl3d */
          if (strchr(line, ','))
            nread = sscanf(line, "%d %d %d %f, %f, %f", &ix, &iz, &iy, &dx, &dz, &dy);
        else
          nread = sscanf(line, "%d %d %d %f %f %f %f %f", &ix, &iz, &iy, &dx,
                         &dz, &dy, &value, &value2);
          if (noflip) {
            itmp = iy;
            iy = iz;
            iz = itmp;
            tmp =dy;
            dy = dz;
            dz = tmp;
          }
          xx = ix;
          yy = iy;
        }
      }
      pts[0].x = xx;
      pts[0].y = yy;
      pts[0].z = iz;
      pts[1].x = xx + scale * dx;
      pts[1].y = yy + scale * dy;
      pts[1].z = iz + scale * dz;
      xmin = B3DMIN(xmin, xx);
      ymin = B3DMIN(ymin, yy);
      zmin = B3DMIN(zmin, iz);
      xmax = B3DMAX(xmax, xx + 1.);
      ymax = B3DMAX(ymax, yy + 1.);
      zmax = B3DMAX(zmax, iz + 1.);
      if (dz != 0.)
        dzvary = 1;

      if (nread > 6) {
        valmin = B3DMIN(valmin, value);
        valmax = B3DMAX(valmax, value);
        if (value || !ignoreZero) {
          nvals++;
          valsum += value;
          sumsq += value * value;
        }
        store.index.i = i;
        store.value.f = value;
        if (istoreInsert(&obj->store, &store))
          exitError("Could not add general storage item");
      }
      if (nread > 7) {
        store2.index.i = i;
        store2.value.f = value2;
        if (istoreInsert(&obj->store, &store2))
          exitError("Could not add general storage item");
      }
    }
    contBase += npatch;
  }
     
  if (residuals) {
    obj->symflags = IOBJ_SYMF_ARROW;
    obj->symbol = IOBJ_SYM_NONE;
    obj->symsize = 7;
  } else if (clipSize) {
    imodViewModelNew(mod);
    for (i = 0; i <  mod->viewsize; i++)
      mod->view[i].world |= WORLD_MOVE_ALL_CLIP;
    obj->clips.count = 4;
    obj->clips.flags = 0;
    obj->clips.normal[0].x = 1.;
    obj->clips.normal[1].x = -1.;
    obj->clips.normal[2].x = 0.;
    obj->clips.normal[3].x = 0.;
    obj->clips.normal[0].y = 0.;
    obj->clips.normal[1].y = 0.;
    obj->clips.normal[2].y = 1.;
    obj->clips.normal[3].y = -1.;
    for (i = 0; i < 4; i++) {
      obj->clips.normal[i].z = 0.;
      obj->clips.point[i].x = -0.5 * (xmax + xmin);
      obj->clips.point[i].y = -0.5 * (ymax + ymin);
      obj->clips.point[i].z = -0.5 * (zmax + zmin);
      if (obj->clips.normal[i].x)
        obj->clips.point[i].x += obj->clips.normal[i].x * clipSize / 2;
      if (obj->clips.normal[i].y)
        obj->clips.point[i].y += obj->clips.normal[i].y * clipSize / 2;
    }      
  }

  mod->xmax = xmax + xmin;
  mod->ymax = ymax + ymin;
  mod->zmax = zmax + zmin;
  if (dzvary)
    obj->symbol = IOBJ_SYM_CIRCLE;

  /* Set current thicken contour flag to aid deleting patches */
  obj->flags |= IMOD_OBJFLAG_THICK_CONT | IMOD_OBJFLAG_MCOLOR;
  if (displayValues)
    obj->flags |= IMOD_OBJFLAG_USE_VALUE;
  if (nvals) {
    if (nvals > 10) {
      valsd = (sumsq - valsum * valsum / nvals) / (nvals - 1);
      if (valsd > 0) {
        sdmax = valsum / nvals + 10. * sqrt(valsd);
        valmax = B3DMIN(valmax, sdmax);
      }
    }

    if (istoreAddMinMax(&obj->store, GEN_STORE_MINMAX1, valmin, valmax))
      exitError("Could not add general storage item");
  }

  if (name)
    imodObjectSetName(obj, name);

  B3DFREE(dxGrid);
  B3DFREE(dyGrid);
  return(mod);
     
}
